DSpace/DSpace
# DS-3940 SHERPA v2 API (main port)

Developed by The Library Code with the support of Technische Universität Hamburg TUHH.

## References
* JIRA link: [DS-3940](https://jira.lyrasis.org/browse/DS-3940)
* SHERPA API key registration: https://v2.sherpa.ac.uk/romeo/api.html
* DSpace 6 PR: #2739 
* The [REST Contract](https://github.com/DSpace/Rest7Contract) entry for this is [External Authority Services](https://github.com/DSpace/Rest7Contract/blob/master/external-authority-sources.md), and is unchanged (though, neither SHERPA source is used as an example in this doc -- it could be added if requested)
* No Angular changes are made as part of this work, it is a refactor of external sources

## Description
This is a master port of #2739 to introduce new SHERPA v2 API data models and integrations, and to refactor existing work in external source providers (for journal and publisher lookups) to make use of these new classes. No Angular changes or REST changes are made. 
It is not an exact port of 6.x as the XMLUI and JSPUI changes are irrelevant, tests have been added, and the default API key is not commented out in dspace.cfg but is instead left blank (as it is required as a placeholder by spring bean configuration). Everything else should be the same (the DSpace 6 version was updated to use DSpace 7 code style rules).

## Instructions for Reviewers
List of changes in this PR:
* Removed old SHERPA model and integration classes and replaced with classes to work with [SHERPA v2 API](https://v2.sherpa.ac.uk/api/object-retrieval.html) 
* Refactored and renamed SherpaJournalDataProvider and SherpaPublisherDataProvider provider classes, to work with the new [SHERPA v2 API](https://v2.sherpa.ac.uk/api/object-retrieval.html) 
* Wrote two unit tests (one for Publisher, one for Journal) to ensure that JSON is parsed correctly. The JSON is taken from raw API responses and supplied as static files in the tests resources bundle

To test this PR:
* [Register for a SHERPA API v2 key](https://v2.sherpa.ac.uk/romeo/api.html) and configure it in local.cfg: **sherpa.romeo.apikey**
* Visit the external sources page in the rest HAL explorer: _http://${server.url}/server/api/integration/externalsources_
* The external sources **sherpaJournal** and **sherpaPublisher** should be listed (as they were previously with the v1 API code)
* For the sherpaJournal source, a search can be tested with a query for a journal name, eg. "PLOS" should return ~7 results.
* For the sherpaPublisher source, a search can be tested with a query for a publisher name, eg. "New Zealand" should return ~15 results
* The "get external object" query (exact title match) can be tested with eg. _http://${server.url}/server/api/integration/externalsources/sherpaJournal/entryValues/PLoS%20Computational%20Biology_

## Checklist
(***My PR is longer, at around ~2000 including comments so probably ~1800-1900 without, but a lot of these changes seem longer since the refactor involves lots of files being replaced / renamed. The changes are isolated just to SHERPA integration and do not impact any other part of the DSpace codebase, so I am hoping this work can still be approved for review***)

- [x] My PR passes all tests and includes new/updated Unit or Integration Tests for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types, including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for known error scenarios and error codes (e.g. `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`, etc)
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
(*It does not include any new dependencies*)
(*It does not modify the REST API, but I have linked to the relevant contract page*)
